### PR TITLE
Fix modern itk and vtk conformance

### PR DESCRIPTION
**Pull request by hjmjohnson**
Update to include conformance with newer ITK and VTK enforced constraints.

ITK 5.3 and VTK 9.1

These changes should be backward compatible.

### CODE REVIEW

**Comment by tashrifbillah** [comment]
Hi Steve (@pieper), will Hans' changes be complatible with Slicer?

**Comment by pieper** [comment]
> Hi Steve (@pieper), will Hans' changes be complatible with Slicer?

Hi @tashrifbillah - I think yes, they look minor and Slicer is pretty up to date with the latest VTK and ITK versions inn prep for Slicer5 release.

**Comment by tashrifbillah** [comment]
Hello @hjmjohnson , did you test as:

```
mkdir build
cd build
cmake ..
make
UKFTractography --help
```

**Comment by hjmjohnson** [comment]
@tashrifbillah   UKF has some tests for building for 8 year old versions of Mac.  Slicer now has a minimum requirement of 10.13 or greater for mac.  Is it OK to update UKF to also have the same minimum requirements as slicer?

PS:  I'm running the build test suggested above, but I am 99% sure it will succeed.

**Comment by tashrifbillah** [comment]
> Is it OK to update UKF to also have the same minimum requirements as slicer?

@pieper , can you answer this quickly? Otherwise, I'll have to do my research on it.

@hjmjohnson , if you are testing on MAC, I can test your PR on CentOS 7 Linux.

@yrathi , do we need support of CentOS 6 anymore?

**Comment by hjmjohnson** [comment]
@tashrifbillah   The mac build is failing due to known problems with the old version of Eigen.  If I update Eigen as well, then the build succeeds.

**Comment by yrathi** [comment]
Thanks @hjmjohnson and @tashrifbillah. We dont need support for CentOS 6. And Yes, we can update UKF to have the minimum requirement as Slicer.

**Comment by hjmjohnson** [comment]
This now builds on mac 10.15 again.  super build packages brought up-to-date with slicer versions of packages.

**Comment by hjmjohnson** [comment]
@yrathi 

I have a series of code maintenance and C++14 performance updates that I would like to apply.  Is there support/interest in adopting coding styles and performance improvements as vetted and incorporated in the ITK and VTK communities for UKF?

The proposed code transformations have been very safe as applied to many ITK-derived projects so far.

**Comment by tashrifbillah** [comment]
Your initial PR failed to build in my CentOS 7 ...
I can test the modification again.

**Comment by hjmjohnson** [comment]
@tashrifbillah NOTE:  Centos7 has a default compiler install of "GCC 4.8.5" which does not fully support C++11.   Centos7 was first released in 2009.

We have a large cluster of computers still running and performing well with Centos7, but they have provided alternative compilers to keep up with code bases that require C++11.

**Comment by hjmjohnson** [comment]
@tashrifbillah Can you let me know what compiler toolchain you were attempting to use on Centos7, and what the error you came across was?

**Comment by tashrifbillah** [comment]
What GCC did you use?

**Comment by hjmjohnson** [comment]
@tashrifbillah my cluster has 8.3.0 available.  I am trying to find the build errors with 4.8.5 now, I expect this build to fail.  I will then attempt a gcc 8.3.0 build.

Hans

**Comment by tashrifbillah** [comment]
I have other GCCs I can try. I just wanted to know what you used. No need to try with CentOS 7 default.

**Comment by yrathi** [comment]
@tashrifbillah -- We should be sure that UKF builds on both the PNL network and the cluster (without having to use any containers etc.). If it doesn't, then we need to think if this update should be applied or not.

**Comment by hjmjohnson** [comment]
@tashrifbillah   I messed up the VTK clone command in a previous patch fix.  The error was not identified until I did a clean build.  updated PR coming soon.

**Comment by hjmjohnson** [comment]
@tashrifbillah NOTE:  Minimum compiler version supported by ITK for C++14 is claimed to be gcc 5.1.  I've added a file to test known minimum compiler versions.  Doing a clean build with gcc8.3.0 ...  SUCCEEDED!

**Comment by hjmjohnson** [comment]
@tashrifbillah Do you have any concerns with this PR?  I would love to get this PR included.

Hans

**Comment by tashrifbillah** [comment]
Sorry Hans, we didn't get time to work on it again yet. I shall get to it soon.

**Comment by tashrifbillah** [comment]
Hans, what is your cmake version?

**Comment by hjmjohnson** [comment]
@tashrifbillah  ITK requires cmake version 3.16.3 to compile.

**Comment by tashrifbillah** [comment]
Testing update:

* builds on CentOS 7 cluster with `cmake-3.19.2` and `GCC-7.3.0`

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,25 +1,38 @@
...
+# to a value greater than the oldest policies, all policies between
+# OLDEST_VALIDATED_POLICIES_VERSION and CMAKE_VERSION (used for this build)
+# are set to their NEW behaivor, thereby suppressing policy warnings related to policies
+# between the OLDEST_VALIDATED_POLICIES_VERSION and CMAKE_VERSION.
+#
+# CMake versions greater than the NEWEST_VALIDATED_POLICIES_VERSION policies will
+# continue to generate policy warnings "CMake Warning (dev)...Policy CMP0XXX is not set:"
+#
+set(OLDEST_VALIDATED_POLICIES_VERSION "3.16.3")
+set(NEWEST_VALIDATED_POLICIES_VERSION "3.19.7")
```
Comment (tashrifbillah) [inline CMakeLists.txt:14]: I used 3.23.0 but I don't think I got the above warning. So would it be okay to have higher cmake than listed?

**Comment by tashrifbillah** [comment]
Now that we are done with building, I shall review your proposals on Eigen, Boost, Apple, and Win32 later.

**Comment by hjmjohnson** [comment]
@tashrifbillah I rebased on the latest master branch.  Is there something I can do to help push this set of changes through?

**Comment by hjmjohnson** [comment]
@tashrifbillah   Are there outstanding concerns with this update?

**Comment by tashrifbillah** [comment]
Sorry we kind of forgot about it. I shall take one final look at it tomorrow and reach out with questions should I have any. Thank you Hans.

**Comment by hjmjohnson** [comment]
@tashrifbillah Any concerns found?

**Comment by hjmjohnson** [comment]
@yrathi Are there outstanding concerns with this PR?  I'm trying to clear my summer todo list.

**Comment by yrathi** [comment]
@hjmjohnson -- I dont think @tashrifbillah  got a chance to double check things. I will wait for his response.

**Comment by hjmjohnson** [comment]
@yrathi Can this be merged yet?

**Comment by tashrifbillah** [comment]
> @tashrifbillah -- We should be sure that UKF builds on both the PNL network and the cluster (without having to use any containers etc.). If it doesn't, then we need to think if this update should be applied or not.

Hans, I see that Jean added more commits. So I need to test this PR again in both places as Yogesh noted.

**Comment by tashrifbillah** [comment]
> Testing update:
> 
>     * builds on CentOS 7 cluster with `cmake-3.19.2` and `GCC-7.3.0`
> 
>     * builds on CentOS 7 workstation with `3.23.0-rc3` and `GCC-7.3.0`

To make GCC-7.3.0 discoverable to `cmake` in workstation:

```
cd ukftractography
mkdir build
cd build

export LD_LIBRARY_PATH=/path/to/gcc-7.3.0/gccbuild/lib/gcc/x86_64-pc-linux-gnu/7.3.0:/path/to/gcc-7.3.0/gccbuild/lib/gcc/x86_64-pc-linux-gnu/7.3.0/lib:/path/to/gcc-7.3.0/gccbuild/lib/gcc/x86_64-pc-linux-gnu/7.3.0/lib64
export PATH=/path/to/gcc-7.3.0/gccbuild/bin:$PATH

CC=/path/to/gcc-7.3.0/gccbuild/bin/gcc CXX=/path/to/gcc-7.3.0/gccbuild/bin/g++ cmake .
```

Note: CC and CXX settings were discovered later. Maybe PATH and LD_LIBRARY_PATH settings are not required. Just PATH and LD_LIBRARY_PATH settings do not make it discoverable.

**Comment by tashrifbillah** [comment]
In workstation, it fails with a different error:
```
> cp b2 bjam
tools/build/src/engine/b2
Detecting Python version... 2.7
Detecting Python root... /usr
Unicode/ICU support for Boost.Regex?... /usr
Generating B2 configuration in project-config.jam for gcc...

Bootstrapping is done. To build, run:

    ./b2
    
To generate header files, run:

    ./b2 headers

The configuration generated uses gcc to build by default. If that is
unintended either use the --with-toolset option or adjust configuration, by
editing 'project-config.jam'.

Further information:

   - Command line help:
     ./b2 --help
     
   - Getting started guide: 
     http://www.boost.org/more/getting_started/unix-variants.html
     
   - B2 documentation:
     http://www.boost.org/build/

[  8%] Performing build step for 'Boost'
./b2: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.20' not found (required by ./b2)
./b2: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.21' not found (required by ./b2)
make[2]: *** [Boost-prefix/src/Boost-stamp/Boost-build] Error 1
make[1]: *** [CMakeFiles/Boost.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by hjmjohnson** [comment]
@tashrifbillah I don't know how to interpret your last messages.  Are you rejecting these changes?  Is the bug related to Jean's changes?  None of the recent messages seem to be related to the PR.

**Comment by tashrifbillah** [comment]
> Is the bug related to Jean's changes?

Yes.

**Comment by tashrifbillah** [comment]
> None of the recent messages seem to be related to the PR.

All of the.

**Comment by tashrifbillah** [comment]
> I don't know how to interpret your last messages.

You can undo Jean's changes or have him fix the errors I reported in https://github.com/pnlbwh/ukftractography/pull/143#issuecomment-1499342115 and https://github.com/pnlbwh/ukftractography/pull/143#issuecomment-1499346856 .

**Comment by hjmjohnson** [comment]
@tashrifbillah I'm going to close this issue as there does not seem to be interest in these updates.

**Pull request closed without merging**